src/main/java/blue/lapis/pore/impl/PoreBanList.java

```diff
@@ -24,38 +24,73 @@
  */
 package blue.lapis.pore.impl;
 
-import org.apache.commons.lang3.NotImplementedException;
+import blue.lapis.pore.Pore;
+import com.google.common.base.Throwables;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import org.bukkit.BanEntry;
 import org.bukkit.BanList;
+import org.spongepowered.api.service.ban.BanService;
+import org.spongepowered.api.util.ban.Ban;
 
+import java.util.Collection;
 import java.util.Date;
+import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
 
-//TODO: skeleton implementation
-public class PoreBanList implements BanList {
+public abstract class PoreBanList<T extends Ban> implements BanList {
+
+    private final Cache<String, PoreBanEntry> cache = CacheBuilder.newBuilder().weakValues().build();
 
     @Override
     public BanEntry getBanEntry(String target) {
-        throw new NotImplementedException("TODO");
+        return getBanFor(target).map(this::loadBanEntry).orElse(null);
     }
 
     @Override
     public BanEntry addBan(String target, String reason, Date expires, String source) {
-        throw new NotImplementedException("TODO");
+        try {
+            BanEntry entry = cache.get(target, () -> new PoreBanEntry(target, this));
```

Comment:
You should use `getUnchecked` here (if available, not sure if that works for the method you can use with the supplier)